Google.Api.Gax.Grpc/Rest/RestMethod.cs

```diff
@@ -82,19 +78,124 @@ internal static RestMethod Create(MethodDescriptor method, JsonParser parser)
+                Content = transcodingResult.Body is null ? null : new StringContent(transcodingResult.Body, Encoding.UTF8, s_applicationJsonMediaType),
             };
         }
 
-        // TODO: Handle cancellation?
+        /// <summary>
+        /// Merges the uri path and the query string parameters, escaping them.
+        /// Ignores the possibility that the path can already have parameters or contain an anchor (`#`)
+        /// </summary>
+        /// <param name="uriPath">The path component of the service URI</param>
+        /// <param name="queryStringParameters">The parameters to encode in the query string</param>
+        /// <returns></returns>
+        private static string AppendQueryStringParameters(string uriPath, IOrderedEnumerable<KeyValuePair<string, string>> queryStringParameters)
+        {
+            var sb = new StringBuilder();
+            sb.Append(uriPath);
+            queryStringParameters.Select((kvpNameValue, i) => new
+            {
+                Name = Uri.EscapeDataString(kvpNameValue.Key),
+                Value = Uri.EscapeDataString(kvpNameValue.Value),
+                Separator = i == 0 ? '?' : '&'
+            }).Aggregate(sb, (sbUri, encodedParam) => sbUri.AppendFormat("{0}{1}={2}", encodedParam.Separator, encodedParam.Name, encodedParam.Value));
+
+            return sb.ToString();
+        }
 
+        /// <summary>
+        /// GRPC Transcoding returns the parameters of the http request, namely
+        ///  - the path of the uri
+        ///  - the query string parameters
+        ///  - the body
+        /// based on the
+        ///  - method descriptor, specifically the list of input message fields, and the `google.api.http` annotation
+        ///  - the request message
+        /// It is described in the AIP-127 (https://google.aip.dev/127) and the proto comments for google.api.HttpRule
+        /// (https://github.com/googleapis/googleapis/blob/master/google/api/http.proto#L44-L312)
+        /// </summary>
+        /// <param name="protoRequest">the request message</param>
+        /// <returns>GRPC Transcoding output</returns>
+        private TranscodingOutput Transcode(IMessage protoRequest)
+        {
+            string path = _pathPattern.Format(protoRequest);
+            (bool allFieldsInBody, string bodyName, string bodyResult) = _httpBody switch
+            {
+                "*" => (true, null, protoRequest.ToString()),
+                "" => (false, null, null),
+                string name when _protoMethod.InputType.FindFieldByName(name) is FieldDescriptor field => (false, name, field.Accessor.GetValue(protoRequest).ToString()),
+                _ => throw new ArgumentException($"Method {_protoMethod.Name} in service {_protoMethod.Service.Name} has a body parameter {_httpBody} in the 'google.api.http' annotation which is not a field in {_protoMethod.InputType.Name}")
+            };
+            
+            var queryStringParams = new Dictionary<string, string>();
```

Comment:
If we use null here instead of creating an empty dictionary on each call, that would be easy to detect in the CreateRequest code.

Alternatively, is there any reason why we need to split this parameter discovery part from the actual append part? Perhaps the TranscodingOutput could just have "path including parameters"?

Reply:
The reason I did it this way is separation of concerns -- I don't want the transcoding part know anything about building the combined URI

Reply:
Hmm... given that it already knows about the base URI and the query parameters themselves, I don't think it's much of a violation to include the encoding here. (After all, it knows to encode the body as JSON... URL-encoding the field value doesn't seem overly different to me.)

Reply:
I think that with the recent changes rollbacking to the contentFactory it's more visible that the transcoding part of the logic doesn't want any more complexity. It can still be done technically. 
(also, I really don't like that transcoding method *knows* that body and query string param names should be JSON but fixin that would invite more ugliness)

Reply:
We can definitely revisit later on.